src/cljam/io/twobit/reader.clj

```diff
@@ -1,125 +1,133 @@
+  (let [floor (or (.floorKey masks (int start)) (int 1))]
+    (doseq [[^long m-start ^long m-size] (.subMap masks floor (int (inc end)))]
+      (when-not (or (< end m-start) (< (+ m-start m-size -1) start))
+        (.position cb (max 0 (- m-start start)))
+        (.mark cb)
+        (let [ca (char-array
+                  (- (min end (+ m-start m-size -1))
+                     (max start m-start) -1))]
+          (.get cb ca)
+          (.reset cb)
+          (dotimes [i (alength ca)]
+            ;; to lower case character
+            (.put cb (unchecked-char
+                      (unchecked-add-int
+                       (unchecked-int (aget ca i)) 32)))))))))
 
 (defn ^String read-sequence
   "Reads sequence at the given region from reader.
    Pass {:mask? true} to enable masking of sequence."
   ([rdr region]
    (read-sequence rdr region {}))
   ([^TwoBitReader rdr {:keys [chr start end]} {:keys [mask?] :or {mask? false}}]
-   (when-let [[n {:keys [offset]}]
-              (first (filter (fn [[i {:keys [name]}]] (= name chr)) (map vector (range) (.file-index rdr))))]
-     (let [{:keys [len ambs masks header-offset]} @(nth (.seq-index rdr) n) ;; Potential seek & read.
-           start' (max 1 (or start 1))
-           end' (min len (or end len))]
+   (when-let [^Chrom c (get (.index rdr) chr)]
+     (let [start' (max 1 (or start 1))
+           end' (min (.len c) (or end (.len c)))]
        (when (<= start' end')
-         (let [start-offset (quot (dec start') 4)
+         ;; Potential seek & read.
+         (let [^ChromHeader h @(.header c)
+               start-offset (quot (dec start') 4)
                end-offset (quot (dec end') 4)
-               ba (byte-array (- end-offset start-offset -1))
-               cb (CharBuffer/allocate (inc (- end' start')))]
-           (.seek ^RandomAccessFile (.reader rdr) (+ offset header-offset start-offset))
-           (.readFully ^RandomAccessFile (.reader rdr) ba)
-           (dotimes [out-pos (inc (- end' start'))]
-             (let [ref-pos (+ out-pos start')
-                   ba-pos (- (quot (dec ref-pos) 4) start-offset)
-                   bit-pos (mod (dec ref-pos) 4)]
-               (when (<= 1 ref-pos len)
-                 (.put cb (.charAt ^String (twobit-to-str (+ (aget ba ba-pos) 128)) bit-pos)))))
-           (replace-ambs! cb ambs start' end')
-           (when mask? (mask! cb masks start' end'))
-           (.rewind cb)
-           (.toString cb)))))))
+               buf ^ByteBuffer (.buf rdr)
+               cb (CharBuffer/allocate (* 4 (inc (- end-offset start-offset))))]
+           (.position buf (+ (.offset c) (.header-offset h) start-offset))
+           (while (.hasRemaining cb)
+             (->> (unchecked-add-int 128 (.get buf))
+                  ^chars (aget twobit-to-str)
+                  (.put cb)))
+           (let [cb' (-> cb
+                         ^CharBuffer (.position (mod (dec start') 4))
```

Comment:
Here's another reflection warning. It can be resolved by coercing (or type-hinting) the `mod` call, or by replacing `mod` with `rem`. I think `rem` is preferable to `mod` whenever both of the arguments are nonnegative because `rem` can be inlined and could be compiled to a JVM instruction directly.